### PR TITLE
Blog: Bringing Software Development practices to Node-RED

### LINKED ISSUE #1497: Blog: Nick (7th Feb)

### PR DESCRIPTION
**Pull request by knolleary**
Closes #1497

## Description

Blog post about bringing common software dev practices to Node-RED. Talks about linting, debugging and testing.

This is an initial draft for early review and feedback.

Still to add:
 - images
 - links to referenced things

### CODE REVIEW

File: src/blog/2024/02/software-development-in-node-red.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+It can help identify potential infinite loops in flows, and highlight Debug nodes without a name set.
+
+
+### Debugging
+
+Whether you are writing lines of code or not, eventually you will need to figure out why your application isn't doing what you think it should.
+
+In Software Engineering there are two typical approaches. One is to add debug statements through the code to print out bits of information as the program runs. Then, depending on what output you got, you'd move the debug statements around, add some more, print out different bits of information - all until you'd nailed down the problem. This is the Debug node approach in Node-RED; adding nodes at different points of your flow to capture some piece of information and then iterating as you go.
+
+This is probably how most Node-RED users go about it today. The downside is you end up leaving the Debug nodes in place, capturing information long after its needed.
```
Comment (gdziuba) [inline src/blog/2024/02/software-development-in-node-red.md:40]:
```suggestion
This is probably how most Node-RED users go about it today. The downside is you end up leaving the Debug nodes in place, capturing information long after it is needed.
```

File: src/blog/2024/02/software-development-in-node-red.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+This allows you to set 'breakpoints' on any node input or output that are triggered when a message arrives at that point of the flow. The Debugger will then pause the whole runtime and shows you all of the queued up messages in the flow.
+
+You can then examine those messages and tell the Debugger to 'release' them one at a time - seeing how the flow progresses.
+
+
+
+### Testing
+
+Testing code is a critical part of developing software. You want to make sure it does what you want. But its more than just manually testing it once and then letting it go; you want to have tests you can run regularly, whenever you make changes, to ensure you don't break something that was working previously.
```
Comment (gdziuba) [inline src/blog/2024/02/software-development-in-node-red.md:52]:
```suggestion
Testing code is a critical part of developing software. You want to make sure it does what you want. But it's more than just manually testing it once and then letting it go; you want to have tests you can run regularly, whenever you make changes, to ensure you don't break something that was working previously.
```

File: src/blog/2024/02/software-development-in-node-red.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+In the Software Development world, there are all sorts of testing methodologies and techniques; unit testing individual components, system testing larger sections, stubbing out components to simulate different conditions, integrating test suites into the whole development process.
+
+They each have their own place in the process of software development. The question is how does this apply to Node-RED?
```
Comment (gdziuba) [inline src/blog/2024/02/software-development-in-node-red.md:56]:
```suggestion
They each have their own place in the process of software development. The question is, how does this apply to Node-RED? 
```

File: src/blog/2024/02/software-development-in-node-red.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+### Debugging
+
+Whether you are writing lines of code or not, eventually you will need to figure out why your application isn't doing what you think it should.
+
+In Software Engineering there are two typical approaches. One is to add debug statements through the code to print out bits of information as the program runs. Then, depending on what output you got, you'd move the debug statements around, add some more, print out different bits of information - all until you'd nailed down the problem. This is the Debug node approach in Node-RED; adding nodes at different points of your flow to capture some piece of information and then iterating as you go.
+
+This is probably how most Node-RED users go about it today. The downside is you end up leaving the Debug nodes in place, capturing information long after its needed.
+
+The alternative approach is Step-by-Step debugging. This is why you are able to pause the program and then step it forward one statement at a time - examining the state at each point. But what's the equivalent for low-code? Pretty much exactly that when you have the Node-RED Debugger plugin installed.
```
Comment (gdziuba) [inline src/blog/2024/02/software-development-in-node-red.md:42]: Link to debugger plug-in?  News to me, and I want to see this.

**Review by gdziuba — COMMENTED** [review]
Thanks Nick.  Looks good.  There were a few things I adjusted, if this is British English vs American English feel free to omit.

**Comment by knolleary** [comment]
Updated with links and images. Will check the deploy preview, but I think this is good for final review and scheduling to be published.

**Review by gdziuba — APPROVED** [review]
Thanks Nick.  I like it.  I am going to start using the debugger.

**Pull request merged**